diffusion/models/layers.py

```diff
@@ -222,3 +223,208 @@ def __call__(
+        q, k, v = self._reshape_for_attention(q), self._reshape_for_attention(k), self._reshape_for_attention(v)
+        h = F.scaled_dot_product_attention(q, k, v, dropout_p=self.dropout_probability)
+        # Reshape back into an image style tensor
+        h = self._reshape_from_attention(h, H, W)
+        # Project to the output
+        h = self.proj_conv(h)
+        return h
+
+
+class Downsample(nn.Module):
+    """Downsampling layer that downsamples by a factor of 2.
+
+    Args:
+        input_channels (int): Number of input channels.
+        resample_with_conv (bool): Whether to use a conv for downsampling.
+    """
+
+    def __init__(self, input_channels: int, resample_with_conv: bool):
+        super().__init__()
+        self.input_channels = input_channels
+        self.resample_with_conv = resample_with_conv
+        if self.resample_with_conv:
+            self.conv = nn.Conv2d(self.input_channels, self.input_channels, kernel_size=3, stride=2, padding=0)
+            nn.init.kaiming_normal_(self.conv.weight, nonlinearity='linear')
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if self.resample_with_conv:
+            # Need to do asymmetric padding to ensure the correct pixels are used in the downsampling conv
+            # and ensure the correct output size is generated for even sizes.
+            x = F.pad(x, (0, 1, 0, 1), mode='constant', value=0)
+            x = self.conv(x)
+        else:
+            x = F.avg_pool2d(x, kernel_size=2, stride=2)
+        return x
+
+
+class Upsample(nn.Module):
+    """Upsampling layer that upsamples by a factor of 2.
+
+    Args:
+        input_channels (int): Number of input channels.
+        resample_with_conv (bool): Whether to use a conv for upsampling.
+    """
+
+    def __init__(self, input_channels: int, resample_with_conv: bool):
+        super().__init__()
+        self.input_channels = input_channels
+        self.resample_with_conv = resample_with_conv
+        if self.resample_with_conv:
+            self.conv = nn.Conv2d(self.input_channels, self.input_channels, kernel_size=3, stride=1, padding=1)
+            nn.init.kaiming_normal_(self.conv.weight, nonlinearity='linear')
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.interpolate(x, scale_factor=2, mode='nearest')
+        if self.resample_with_conv:
+            x = self.conv(x)
+        return x
+
+
+class GradientScalingLayer(nn.Module):
+    """Layer that scales the gradient by a multiplicative constant.
+
+    By default, this constant is 1.0, so this layer acts as an identity function.
+
+    To use, one must also register the backward hook:
+    scaling_layer = GradientScalingLayer()
+    scaling_layer.register_full_backward_hook(scaling_layer.backward_hook)
+
+    And then set the scale via
+    scaling_layer.set_scale(scale)
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.scale = 1.0
```

Comment:
You could reduce some of this boilerplat by having a dataclass.
```suggestion
        self.scale : float = 1.0
```

Reply:
Added the type here. Can you elaborate on the point about reducing the boilerplate?